src/modules/presentation-exchange/services/vc-match.service.ts

```diff
@@ -0,0 +1,75 @@
+import { Injectable } from '@nestjs/common';
+import { PEX } from '@sphereon/pex';
+import { Logger } from '../../logger/logger.service';
+import { HandlerService } from '../../decentralized-web-node/services/handler.service';
+import { MatchResultsDto } from '../dtos/match-results.dto';
+import { PresentationDefinitionDto } from '../dtos/presentation-definition.dto';
+import {
+  VerifiableCredential,
+  isVerifiableCredential,
+} from '@ew-did-registry/credentials-interface';
+import { MessageDto } from '../../decentralized-web-node/dtos/message.dto';
+import { CollectionsQueryDescriptorDto } from '../../decentralized-web-node/dtos/collections-query-descriptor.dto';
+
+@Injectable()
+export class VCMatchService {
+  constructor(
+    private readonly logger: Logger,
+    private readonly handlerService: HandlerService
+  ) {
+    this.logger.setContext(VCMatchService.name);
+  }
+
+  private decodeData(
+    writeMessage: unknown
+  ): Record<string, unknown> | undefined {
+    try {
+      if (!writeMessage['data']) return;
+      const dataBase64 = Buffer.from(writeMessage['data'], 'base64').toString(
+        'utf8'
+      );
+      return JSON.parse(dataBase64);
+    } catch {
+      return;
+    }
+  }
+
+  private async queryHolderVCs(
+    holder: string
+  ): Promise<VerifiableCredential<Record<string, unknown>>[]> {
+    const queryMessage: MessageDto & {
+      descriptor: CollectionsQueryDescriptorDto;
+    } = {
+      descriptor: {
+        method: 'CollectionsQuery',
+        schema: 'https://www.w3.org/TR/vc-data-model/',
+        dataFormat: 'application/ld+json',
+      },
+    };
+
+    const reply = await this.handlerService.handleMessage(queryMessage, holder);
+
+    if (reply.status.code !== 200 || !Array.isArray(reply.entries)) {
+      throw new Error(
+        `Failed to query holder ${holder} for verifiable credentials`
+      );
+    }
+
+    const entries = reply.entries
+      .map(this.decodeData)
+      .filter(Boolean)
+      .filter(isVerifiableCredential);
+
+    return entries;
+  }
+
+  async matchCredentials(
```

Comment:
Minor style point, but IMO, it's better to put public members of a class at the top as a rule of thumb. This is because it helps readers find the interface of the class quickly without having to look at implementation details.
Some discussion about this here. There definitely isn't clear agreement on this, but just my 2 cents https://stackoverflow.com/questions/1760850/best-practice-ordering-of-public-protected-private-within-the-class-definition